plutus-core/plutus-ir/src/PlutusIR/Transform/NewLetFloat.hs

```diff
@@ -0,0 +1,493 @@
+   If yes, the let-group's binding is floated exactly outside a lambda abstraction that has lam_depth=let_depth+1
+
+There are some  differences with the paper's described implementation above, namely:
+
+a) we use 3 passes. the 1st pass is similar to the original; a second pass
+"cleans" the term from all the to-be-floated lets and stores them in a separate table.
+the 3rd pass is responsible to float back the removed lets inside the reducted (cleaned) term
+according to their markers. So we use an extra pass because we float back lets in a global fashion,
+instead of deciding locally.
+
+b) Since the 3rd (float-back) pass operates on a reducted (cleaned) term, we have lost
+the original location of the lets, so we cannot float them "right outside" the **maximum-independent lambda-abstraction**,
+but we float them "right inside" the maximum **dependent** lambda-abstraction's body. This has the downside
+of allocating&holding the lets for longer than needed, but will not alter the meaning of the original PIR term.
+
+c) Since PIR has strict (compared to the paper's lazy-only lang), we have to make
+sure that any let-group containing at least one **effectful** (i.e. non-pure) strict binding is
+not floated at all. This does not mean that such an "effectful" let
+will appear in the same absolute location as the original term: an outside/parent let may float around,
+changing the child's (effectful let) absolute location; however, the child's relative location to the parent *must* remain the same.
+See also 'floatable'. Consider this example:
+
+.... let (nonstrict) x1= (let (strict) x2 = error in rhs1) in body1
+
+the parent is x1 and it is floatable
+the child is x2 and is unmovable
+The x2 will not move with respect to x1, but its original absolute program location may change, because the parent may float upwards somewhere else.
+
+Since another let variable may depend on such *effectful* let and to preserve the execution order,
+we treat an effectful let the same as lam/Lam "anchor", by increasing the depth both on entering any of its rhs'es
+*and* inside its inTerm. See also 'PosType'. Thus, the dependent let will be etierh floated right together with a depending-let group or
+right inside its depending-let inTerm.
+-}
+
+-- | Position of a marker.
+-- 1) Since we act globally the depth is not enough anymore (not globally unique)
+-- and for that we also use a "represenative" identifier (PLC.Unique).
+-- 2) Since we have lets as anchors, we also need the extra 'PosType' to signify
+-- the let's marker position (at rhs or at inTerm).
+type Pos = ( Depth
+           , PLC.Unique -- The lambda name or Lambda tyname or Let's representative unique
+           , PosType
+           )
+
+data PosType = LamBody -- ^ big, small lam or let body
+             | LetRhs -- ^ let rhs or top
+             deriving (Eq, Ord, Show)
+
+
+-- Arbitrary: return a single unique among all the introduced uniques of the given letgroup.
+representativeBindingUnique
```

Comment:
If a let group is unfloatable, we treat it as a whole by using a single `Unique` to represent the whole unfloatable let-group position. Of course this means that some floatable lets may originate from different rhs'es of the same let group, but this is 
 this fine, because we place these floatable lets right together with this group. Also, we don't have to change the group's recursivity, because it is implied that the group was originally rec, since we "float up from inside of one of the rhs'es", so it must be rec anyway.

Reply:
I'm worried about putting things back. I guess the thing that I'm worried about is a recursive let-binding (so it won't be broken up), where we need to float things into several different RHSs. i.e. does this work:
```
let rec
   x = (let x' = x in x')
   y = (let y' = y in y')
in x
```
I'm worried that we'd end up with something like
```
let rec
   x = (let x' = x; y' = y in x')
   y = y'
in x
```
because we only have one position to float to?

Reply:
In your example:
```
(let
   (rec)
   (termbind (nonstrict) (vardecl x (con integer))
      (let
         (rec)
         (termbind (nonstrict) (vardecl x' (con integer)) x)
         x')
         )
   (termbind (nonstrict) (vardecl y (con integer))
      (let
         (rec)
         (termbind (nonstrict) (vardecl y' (con integer)) y)
         y')
         )
   x
)
```

, this is  what happens:

x binding is floatable , has no freevars so its floating position x_pos = Top (-1)
y binding is floatable , has no freevars so its floating position y_pos = Top (-1)
x' has maximal free var x so  its floating position is the floating position of x , x'_pos = x_pos
y' has maximal free var x so  its floating position is the floating position of y , y'_pos = y_pos

this means the the cleaned term is just `x` and when floatBackLets.goTop we create a brand new lambda with all of those:

```
(let
  (rec)
  (termbind (nonstrict) (vardecl x' (con integer)) x)
  (termbind (nonstrict) (vardecl y' (con integer)) y)
  (termbind (nonstrict) (vardecl x (con integer)) x')
  (termbind (nonstrict) (vardecl y (con integer)) y')
  x
)
```

this output is without RecSplit.  After applying also RecSplit we get:

```
(let
  (rec)
  (termbind (nonstrict) (vardecl y (con integer)) y')
  (termbind (nonstrict) (vardecl y' (con integer)) y)
  (let
    (rec)
    (termbind (nonstrict) (vardecl x (con integer)) x')
    (termbind (nonstrict) (vardecl x' (con integer)) x)
    x
  )
)
```

Reply:
This particular RecSplit order is because how the x and y appeared in the original program. If we change their order:

```
(let
   (rec)
(termbind (nonstrict) (vardecl y (con integer))
      (let
         (rec)
         (termbind (nonstrict) (vardecl y' (con integer)) y)
         y')
         )   
(termbind (nonstrict) (vardecl x (con integer))
      (let
         (rec)
         (termbind (nonstrict) (vardecl x' (con integer)) x)
         x')
         )
   x
)
```

We get after letfloat+recsplit:

```
(let
  (rec)
  (termbind (nonstrict) (vardecl x (con integer)) x')
  (termbind (nonstrict) (vardecl x' (con integer)) x)
  (let
    (rec)
    (termbind (nonstrict) (vardecl y (con integer)) y')
    (termbind (nonstrict) (vardecl y' (con integer)) y)
    x
  )
)
```

It just a different scc order